### PR TITLE
Move BDA::read into StaticHeader::read

### LINKED ISSUE #1582: Move BDA::read out of BDA implementation
It only reads the locations of the sigblocks and the only method that depends on it is ```StaticHeader::setup```.

~~We should probably make it a member of the static_header module, rather than a member of the ```StaticHeader``` impl, also.~~ We'll handle this part of the big move later if it seems indicated. Let's just move it into the ```StaticHeader``` impl, so that it can be with its friend, ```StaticHeader::write```.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #1582.

### CODE REVIEW

**Comment by mulkieran** [comment]
ok to test.

File: src/engine/strat_engine/backstore/metadata/bda.rs [hunk 1]
```diff
@@ -270,6 +223,53 @@ impl StaticHeader {
         }
     }
 
+    /// Read the BDA from the device and return 2 SECTORS worth of data, one for each BDA returned
```
Comment (mulkieran) [inline src/engine/strat_engine/backstore/metadata/bda.rs]: Please add an additional commit to change the header comment to something like this:

/// Read the data at both signature block locations.
/// Return the data from each location as an array of bytes
/// or an error if the read fails. Return the values
/// in the order in which they occur on the device.
/// Read the contents of each signature block separately,
/// as this increases the probability that at least one read
/// will not fail.

File: src/engine/strat_engine/backstore/metadata/bda.rs [hunk 2]
```diff
@@ -270,6 +223,53 @@ impl StaticHeader {
...
+    fn read<F>(
+        f: &mut F,
+    ) -> (
+        io::Result<[u8; bytes!(static_header_size::SIGBLOCK_SECTORS)]>,
+        io::Result<[u8; bytes!(static_header_size::SIGBLOCK_SECTORS)]>,
+    )
+    where
+        F: Read + Seek,
+    {
+        // Theory of read procedure
```
Comment (mulkieran) [inline src/engine/strat_engine/backstore/metadata/bda.rs]: Please remove this comment, too. The topic is covered in the suggested header comment, and a lot of the nomenclature is wrong in this comment.

File: src/engine/strat_engine/backstore/metadata/bda.rs [hunk 3]
```diff
@@ -270,6 +223,53 @@ impl StaticHeader {
...
+        // We write the BDA in two operations with a sync in between.  The write operation
+        // could fail (loss of power) for either write leaving sector(s) with potentially hard
+        // read errors. It's best to read each of the specific BDA blocks individually, to limit
+        // the probability of hitting a read error on a non-essential sector.
+
+        let mut buf_loc_1 = [0u8; bytes!(static_header_size::SIGBLOCK_SECTORS)];
+        let mut buf_loc_2 = [0u8; bytes!(static_header_size::SIGBLOCK_SECTORS)];
+
+        /// Read a bda sector worth of data at the specified offset into buffer.
```
Comment (mulkieran) [inline src/engine/strat_engine/backstore/metadata/bda.rs]: Might as well omit this header comment as well. It is a bit misleading also.

File: src/engine/strat_engine/backstore/metadata/bda.rs [hunk 4]
```diff
@@ -270,6 +223,53 @@ impl StaticHeader {
...
+        fn read_sector_at_offset<F>(f: &mut F, offset: usize, mut buf: &mut [u8]) -> io::Result<()>
+        where
+            F: Read + Seek,
+        {
+            f.seek(SeekFrom::Start(offset as u64))?;
```
Comment (mulkieran) [inline src/engine/strat_engine/backstore/metadata/bda.rs]: In another separate commit, could you tighten up the body to one expression, as:

```f.seek(SeekFrom::Start(offset as u64)).and_then(|_| f.read_exact(&mut buf))```

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by mulkieran** [comment]
Drat. I accidentally pressed the merge button.

**Comment by GuillaumeGomez** [comment]
Removed merge commit and updated.

**Comment by mulkieran** [comment]
Thanks!

**Comment by mulkieran** [comment]
Travis is just being a bit flaky.

**Pull request merged**